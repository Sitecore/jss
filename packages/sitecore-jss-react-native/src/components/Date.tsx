import React from 'react';
import PropTypes from 'prop-types';
import { Text } from 'react-native'
import HtmlView from 'react-native-htmlview';

export interface DateFieldProps {
	/** The date field data. */
	field: {
		value?: string;
		editable?: string;
	};
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   * @default true
   */
	editable?: boolean;
	render?: (date: Date | null) => React.ComponentType<any> | React.ReactNode;
	[htmlAttributes: string]: any;
}

export const DateField: React.SFC<DateFieldProps> = ({ field, editable, render, ...otherProps }) => {
	if (!field || (!field.editable && !field.value)) {
		return null;
	}

	let children: any;

	if (render) {
		children = render(field.value ? new Date(field.value) : null);
	} else {
		children = field.value
	}

	if (field.editable && editable) {
		return <HtmlView value={children} {...otherProps} />;
	} else {
		return render 
			? children
			: <Text>{children}</Text>;
	}
};

DateField.propTypes = {
	field: PropTypes.shape({
		value: PropTypes.any,
		editable: PropTypes.string,
	}).isRequired,
	editable: PropTypes.bool,
	render: PropTypes.func,
};

DateField.defaultProps = {
	editable: true,
};

DateField.displayName = 'Date';