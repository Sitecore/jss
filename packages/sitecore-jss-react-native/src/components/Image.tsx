import React from 'react';
import PropTypes from 'prop-types';
import { Image as NativeImage, ImageSourcePropType } from 'react-native';
import SvgUri from 'react-native-svg-uri';
import { mediaApi } from '@sitecore-jss/sitecore-jss';

export interface ImageFieldValue {
  /** HTML attributes that will be appended to the rendered <img /> tag. */
  [attributeName: string]: unknown;
  src?: string | number;
  width?: number;
  height?: number;
  style?: unknown;
}

export interface ImageField {
  value?: ImageFieldValue;
  editable?: string;
}

export interface ImageProps {
  /** HTML attributes that will be appended to the rendered <img /> tag. */

  [attributeName: string]: unknown;
  /** The image field data. */
  media: ImageField | ImageFieldValue | null;
  field?: ImageField | ImageFieldValue | null;

  /**
   * Parameters that will be attached to Sitecore media URLs
   */
  imageUrlParams?: {
    [paramName: string]: string;
  } | null;
}

const getImageAttrs = (
  {
    src,
    width,
    height,
    style,
    ...otherAttrs
  }: {
    [attr: string]: unknown;
    src?: string | number;
    width?: number | undefined;
    height?: number | undefined;
    style?: unknown;
  },
  imageUrlParams?: { [paramName: string]: string } | null
) => {
  if (!src) {
    return null;
  }

  const newAttrs: { [key: string]: unknown } = { ...otherAttrs };

  // for network images, "width" and "height" are required as style properties, otherwise the image likely won't display.
  // for static images, "width" and "height" are not required, but if they are passed in then add them to the style prop.
  const imgStyles: { [attr: string]: unknown } = {};
  // react-native doesn't seem to like `&&` assignment, so use if statements instead
  if (width) {
    imgStyles.width = typeof width !== 'number' ? Number(width) : width;
  }
  if (height) {
    imgStyles.height = typeof height !== 'number' ? Number(height) : height;
  }

  // react native styles/StyleSheets can be merged by passing them as an array to the native Image component
  newAttrs.style = style ? [imgStyles, style] : imgStyles;

  // in react-native, you need to "import" static assets via `require` statements
  // when the packager builds your app, it (presumably) statically analyzes your code, extracts
  // the assets that are `require`d into an array/map, then assigns them a numerical value.
  if (typeof src === 'number') {
    newAttrs.source = src;
  } else {
    // assume we have a "network image", i.e. something loaded via http(s)
    // update image URL for jss handler and image rendering params
    const uri = mediaApi.updateImageUrl(src, imageUrlParams);
    // for network images, the "source" prop is an object with at least "uri" value, e.g. { uri: 'http://aviato.com/myimg.jpg' }
    newAttrs.source = { uri };
  }

  return newAttrs;
};

export const isSvgImage = (source: ImageSourcePropType) => {
  const isSvg = /\.svg($|\?|\&)/g;
  const src = NativeImage.resolveAssetSource(source);

  return src && isSvg.test(src.uri);
};

export const Image: React.SFC<ImageProps> = ({ media, imageUrlParams, field, ...otherProps }) => {
  let dynamicMedia = media;

  // allows the mistake of using 'field' prop instead of 'media' (consistent with other helpers)
  if (field && !media) {
    dynamicMedia = field;
  }

  if (!dynamicMedia || (!dynamicMedia.value && !(dynamicMedia as ImageFieldValue).src)) {
    return null;
  }

  // some wise-guy/gal might pass in a 'raw' image object value
  const img = (dynamicMedia as ImageFieldValue).src
    ? dynamicMedia
    : (dynamicMedia as ImageField).value;
  if (!img) {
    return null;
  }

  const attrs = getImageAttrs({ ...(img as ImageFieldValue), ...otherProps }, imageUrlParams);

  if (attrs && isSvgImage(attrs.source as ImageSourcePropType)) return <SvgUri {...attrs} />;

  if (attrs) return <NativeImage {...attrs} source={attrs.source as ImageSourcePropType} />;

  return null; // we can't handle the truth
};

Image.propTypes = {
  media: PropTypes.oneOfType([
    PropTypes.shape({
      src: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    }),
    PropTypes.shape({
      value: PropTypes.object,
    }),
  ]),
  imageUrlParams: PropTypes.any,
};

Image.displayName = 'Image';
