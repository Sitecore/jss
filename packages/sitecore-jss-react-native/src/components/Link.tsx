import React, { ReactElement } from 'react';
import PropTypes from 'prop-types';
import HtmlView from 'react-native-htmlview';
import { Text, View, Linking, TouchableWithoutFeedback } from 'react-native';

export interface LinkFieldValue {
  [attributeName: string]: unknown;
  href?: string;
  text?: string;
}

export interface LinkField {
  value: LinkFieldValue;
  editableFirstPart?: string;
  editableLastPart?: string;
}

export type LinkProps = {
  /** The link field data. */
  field: LinkField | LinkFieldValue;
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   * @default true
   */
  editable?: boolean;

  /**
   * Displays a link text ('description' in Sitecore) even when children exist
   * NOTE: when in Sitecore Experience Editor, this setting is ignored due to technical limitations, and the description is always rendered.
   */
  showLinkTextWithChildrenPresent?: boolean;

  style?: { [attr: string]: unknown };
  textStyle?: { [attr: string]: unknown };
};

export const Link: React.FunctionComponent<LinkProps> = ({
  field,
  editable,
  children,
  showLinkTextWithChildrenPresent,
  style,
  textStyle,
  ...otherProps
}) => {
  const dynamicField = field;

  if (
    !field ||
    (!dynamicField.editableFirstPart &&
      !dynamicField.value &&
      !(dynamicField as LinkFieldValue).href)
  ) {
    return null;
  }

  const resultTags: ReactElement<unknown>[] = [];

  const { editableFirstPart, editableLastPart } = dynamicField as LinkField;

  // EXPERIENCE EDITOR RENDERING
  if (editable && editableFirstPart) {
    const markup: string = editableFirstPart + editableLastPart;

    // in an ideal world, we'd pre-render React children here and inject them between editableFirstPart and editableLastPart.
    // However, we cannot combine arbitrary unparsed HTML (innerHTML) based components with actual vDOM components (the children)
    // because the innerHTML is not parsed - it'd make a discontinuous vDOM. So, we'll go for the next best compromise of rendering the link field and children separately
    // as siblings. Should be "good enough" for most cases - and write your own helper if it isn't. Or bring xEditor out of 2006.

    resultTags.push(<HtmlView value={markup} />);

    // don't render normal link tag when editing, if no children exist
    // this preserves normal-ish behavior if not using a link body (no hacks required)
    if (!children) {
      return resultTags[0];
    }
  }

  // handle link directly on field for forgetful devs
  const link = ((dynamicField as LinkFieldValue).href
    ? field
    : (dynamicField as LinkField).value) as LinkFieldValue;

  if (!link) {
    return null;
  }

  const linkText = showLinkTextWithChildrenPresent || !children ? link.text || link.href : null;

  resultTags.push(
    <TouchableWithoutFeedback
      onPress={() => link.href && Linking.openURL(link.href)}
      key="link"
      {...otherProps}
    >
      <View style={style}>
        {linkText && <Text style={textStyle}>{linkText}</Text>}
        {children}
      </View>
    </TouchableWithoutFeedback>
  );

  return <View>{resultTags}</View>;
};

Link.propTypes = {
  field: PropTypes.oneOfType([
    PropTypes.shape({
      href: PropTypes.string,
    }),
    PropTypes.shape({
      value: PropTypes.object,
      editableFirstPart: PropTypes.string,
      editableLastPart: PropTypes.string,
    }),
  ]).isRequired,
  editable: PropTypes.bool,
};

Link.defaultProps = {
  editable: true,
};

Link.displayName = 'Link';
