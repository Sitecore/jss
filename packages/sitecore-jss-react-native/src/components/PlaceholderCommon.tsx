import React from 'react';
import PropTypes from 'prop-types';
import { MissingComponent } from './MissingComponent';
import { ComponentFactory } from './sharedTypes';
import {
  ComponentRendering,
  RouteData,
  Field,
  Item,
  HtmlElementRendering,
} from '@sitecore-jss/sitecore-jss';
import { UnrenderableComponent } from './UnrenderableComponent';

export interface PlaceholderProps {
  [key: string]: unknown;
  /** Name of the placeholder to render. */
  name: string;
  /** Rendering data to be used when rendering the placeholder. */
  rendering: ComponentRendering | RouteData;
  /**
   * A factory function that will receive a componentName and return an instance of a React component.
   * When rendered within a <SitecoreContext> component, defaults to the context componentFactory.
   */
  componentFactory?: ComponentFactory | null;
  /**
   * An object of field names/values that are aggregated and propagated through the component tree created by a placeholder.
   * Any component or placeholder rendered by a placeholder will have access to this data via `props.fields`.
   */
  fields?: {
    [name: string]: Field | Item[];
  } | null;
  /**
   * An object of rendering parameter names/values that are aggregated and propagated through the component tree created by a placeholder.
   * Any component or placeholder rendered by a placeholder will have access to this data via `props.params`.
   */
  params?: {
    [name: string]: string;
  } | null;

  /**
   * A component that is rendered in place of any components that are in this placeholder,
   * but do not have a definition in the componentFactory (i.e. don't have a React implementation)
   */
  missingComponentComponent?: React.ComponentClass<unknown> | React.SFC<unknown> | null;

  /**
   * A component that is rendered in place of the placeholder when an error occurs rendering
   * the placeholder
   */
  errorComponent?: React.ComponentClass<unknown> | React.SFC<unknown> | null;

  /**
   * A component that is rendered in place of any components that are in this placeholder,
   * but are not renderable by react-native (i.e. DOM elements)
   */
  unrenderableComponentComponent?: React.ComponentClass<unknown> | React.SFC<unknown> | null;
}

export class PlaceholderCommon extends React.Component<PlaceholderProps> {
  static propTypes = {
    name: PropTypes.string.isRequired,
    componentFactory: PropTypes.func,
    rendering: PropTypes.any.isRequired,
    fields: PropTypes.any,
    params: PropTypes.any,
    missingComponentComponent: PropTypes.any,
    errorComponent: PropTypes.any,
    unrenderableComponentComponent: PropTypes.any,
  };

  nodeRefs: unknown[];
  state: Readonly<{ error?: Error }>;

  constructor(props: PlaceholderProps) {
    super(props);
    this.nodeRefs = [];
    this.state = {};
  }

  static getPlaceholderDataFromRenderingData(
    rendering: ComponentRendering | RouteData,
    name: string
  ) {
    let result;
    if (rendering && rendering.placeholders && Object.keys(rendering.placeholders).length > 0) {
      result = rendering.placeholders[name];
    } else {
      result = null;
    }

    if (!result) {
      console.warn(
        `Placeholder '${name}' was not found in the current rendering data`,
        JSON.stringify(rendering, null, 2)
      );

      return [];
    }

    return result;
  }

  componentDidCatch(error: Error) {
    this.setState({ error });
  }

  getComponentsForRenderingData(
    placeholderData: (ComponentRendering | HtmlElementRendering)[] | null
  ) {
    const {
      name,
      fields: placeholderFields,
      params: placeholderParams,
      unrenderableComponentComponent,
      missingComponentComponent,
      ...placeholderProps
    } = this.props;

    return (
      placeholderData &&
      placeholderData
        .map((rendering: ComponentRendering | HtmlElementRendering, index: number) => {
          const componentRendering = rendering as ComponentRendering;
          const htmlElementRendering = rendering as HtmlElementRendering;

          const key = componentRendering.uid ? componentRendering.uid : `component-${index}`;
          const commonProps = { key };

          let component: React.ReactNode | null;
          // if the element is not a 'component rendering', we can't render it 'raw' like with react-dom
          // register a warning instead.
          if (!componentRendering.componentName && htmlElementRendering.name) {
            console.error(
              `Placeholder ${name} contains a rendering that cannot be rendered in React Native '${htmlElementRendering.name}'. This is likely the result of including Experience Editor output in rendering data
            or using non-JSON renderings in an item's presentation details / layout. React Native
            is not able to render DOM elements, your Sitecore renderings must map to React components
            defined in your componentFactory.js.`
            );
            component = unrenderableComponentComponent || UnrenderableComponent;
          }

          if (!component) {
            component = this.getComponentForRendering(componentRendering);
            if (!component) {
              console.error(
                `Placeholder ${name} contains unknown component ${componentRendering.componentName}. Ensure that a React component exists for it, and that it is registered in your componentFactory.js.`
              );

              component = missingComponentComponent || MissingComponent;
            }
          }

          const finalProps = {
            ...commonProps,
            ...placeholderProps,
            ...((placeholderFields || componentRendering.fields) && {
              fields: { ...placeholderFields, ...componentRendering.fields },
            }),
            ...((placeholderParams || componentRendering.params) && {
              params: { ...placeholderParams, ...componentRendering.params },
            }),
            rendering,
          };

          return React.createElement<{ [attr: string]: unknown }>(
            component as React.ComponentType,
            finalProps
          );
        })
        .filter((element: React.ReactNode) => element)
    ); // remove nulls
  }

  getComponentForRendering(renderingDefinition: { componentName: string }) {
    const componentFactory = this.props.componentFactory;

    if (!componentFactory || typeof componentFactory !== 'function') {
      console.warn(
        `No componentFactory was available to service request for component ${renderingDefinition}`
      );
      return null;
    }

    return componentFactory(renderingDefinition.componentName);
  }
}
