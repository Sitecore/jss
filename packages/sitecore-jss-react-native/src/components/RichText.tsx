import React from 'react';
import HtmlView from 'react-native-htmlview';
import PropTypes from 'prop-types';

export interface RichTextProps {
  [htmlViewProps: string]: unknown;
  /** The rich text field data. */
  field: {
    value?: string;
    editable?: string;
  } | null;
}

export const RichText: React.SFC<RichTextProps> = ({ field, ...otherProps }) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  const htmlValue = field.editable ? field.editable : field.value;

  return <HtmlView value={htmlValue as string} {...otherProps} />;
};

RichText.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.string,
    editable: PropTypes.string,
  }),
};

RichText.displayName = 'RichText';
