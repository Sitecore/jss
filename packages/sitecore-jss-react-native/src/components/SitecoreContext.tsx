import React from 'react';
import PropTypes from 'prop-types';
import { ComponentFactory } from './sharedTypes';

export interface SitecoreContextProps {
  componentFactory: ComponentFactory;
  children: unknown;
}

export const ComponentFactoryReactContext = React.createContext<ComponentFactory>(
  {} as ComponentFactory
);

export class SitecoreContext extends React.Component<SitecoreContextProps> {
  static propTypes = {
    children: PropTypes.any.isRequired,
    componentFactory: PropTypes.func,
  };

  static displayName = 'SitecoreContext';

  componentFactory: ComponentFactory;

  constructor(props: SitecoreContextProps, context: unknown) {
    super(props, context);
    this.componentFactory = props.componentFactory;
  }

  render() {
    return (
      <ComponentFactoryReactContext.Provider value={this.componentFactory}>
        {this.props.children}
      </ComponentFactoryReactContext.Provider>
    );
  }
}
