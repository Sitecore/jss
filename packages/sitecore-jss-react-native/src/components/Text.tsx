import React from 'react';
import PropTypes from 'prop-types';
import { Text as NativeText } from 'react-native';

export interface TextProps {
  [nativeTextProps: string]: unknown;
  /** The text field data. */
  field: {
    value?: string;
    editable?: string;
  } | null;
}

export const Text: React.SFC<TextProps> = ({ field, ...otherProps }) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  const textValue = field.editable ? field.editable : field.value;

  return <NativeText {...otherProps}>{textValue}</NativeText>;
};

Text.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.any,
    editable: PropTypes.string,
  }),
};

Text.displayName = 'Text';
