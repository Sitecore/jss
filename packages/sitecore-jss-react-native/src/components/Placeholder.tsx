import React from 'react';
import { View } from 'react-native';
import { withComponentFactory } from '../enhancers/withComponentFactory';
import { ComponentRendering, HtmlElementRendering } from '@sitecore-jss/sitecore-jss';
import { PlaceholderCommon, PlaceholderProps } from './PlaceholderCommon';

export interface PlaceholderComponentProps extends PlaceholderProps {
  /**
   * Render props function that enables control over the rendering of the components in the placeholder.
   * Useful for techniques like wrapping each child in a wrapper component.
   */
  render?: (
    components: React.ReactNode[],
    data: (ComponentRendering | HtmlElementRendering)[],
    props: PlaceholderProps
  ) => React.ComponentClass<unknown> | React.SFC<unknown>;

  /**
   * Render props function that is called when the placeholder contains no content components.
   * Can be used to wrap the Sitecore EE empty placeholder markup in something that's visually correct
   */
  renderEmpty?: (
    components: React.ReactNode[],
    data: (ComponentRendering | HtmlElementRendering)[],
    props: PlaceholderProps
  ) => React.ComponentClass<unknown> | React.SFC<unknown> | React.ReactNode;

  /**
   * Render props function that is called for each non-system component added to the placeholder.
   * Mutually exclusive with `render`. System components added during Experience Editor are automatically rendered as-is.
   */
  renderEach?: (
    component: React.ReactNode,
    index: number
  ) => React.ComponentClass<unknown> | React.SFC<unknown> | React.ReactNode;
}

/**
 * @param {HtmlElementRendering | ComponentRendering} rendering
 */
function isRawRendering(
  rendering: HtmlElementRendering | ComponentRendering
): rendering is HtmlElementRendering {
  return (
    !(rendering as ComponentRendering).componentName &&
    (rendering as HtmlElementRendering).name !== undefined
  );
}

class PlaceholderComponent extends PlaceholderCommon {
  static propTypes = PlaceholderCommon.propTypes;

  constructor(props: PlaceholderComponentProps) {
    super(props);
  }

  render() {
    const props: PlaceholderComponentProps = this.props;
    const childProps: PlaceholderComponentProps = { ...this.props };

    delete childProps.componentFactory;
    delete childProps.render;
    delete childProps.renderEach;

    if (this.state.error) {
      return <View>A rendering error occurred: {this.state.error.message}.</View>;
    }

    const renderingData = childProps.rendering;

    const placeholderData = PlaceholderCommon.getPlaceholderDataFromRenderingData(
      renderingData,
      props.name
    );
    const components = this.getComponentsForRenderingData(placeholderData);

    if (
      props.renderEmpty &&
      placeholderData.every((rendering: ComponentRendering | HtmlElementRendering) =>
        isRawRendering(rendering)
      )
    ) {
      return components && props.renderEmpty(components, placeholderData, childProps);
    } else if (props.render) {
      return components && props.render(components, placeholderData, childProps);
    } else if (props.renderEach) {
      return (
        components &&
        components.map((component, index) => {
          if (component && component.props && component.props.type === 'text/sitecore') {
            return component;
          }

          return props.renderEach && props.renderEach(component, index);
        })
      );
    } else {
      return components;
    }
  }
}

export const Placeholder = withComponentFactory(PlaceholderComponent);
