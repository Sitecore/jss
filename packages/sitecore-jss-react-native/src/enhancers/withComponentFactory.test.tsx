import React from 'react';
import { View, Text } from 'react-native';
import renderer from 'react-test-renderer';
import { withComponentFactory } from './withComponentFactory';
import { ComponentFactory } from '../components/sharedTypes';
import { ComponentFactoryReactContext } from '../components/SitecoreContext';

describe('withComponentFactory()', () => {
  test('should pass factory to wrapped component from context', () => {
		const components = new Map();

    components.set('xxx', () => <Text>I'm component from factory</Text>);

    const componentFactory: ComponentFactory = (componentName) => components.get(componentName);

    type TestComponentProps = {
      customProperty?: string;
      componentFactory?: ComponentFactory | null;
    }

    const TestComponent = (props: TestComponentProps) => {
			if (!props.componentFactory) return null;

      const ComponentFromFactory = props.componentFactory('xxx') as React.ComponentType;

      return (
        <View>
          <Text>Hello world...</Text>
          <Text>Custom property:{props.customProperty}</Text>
          <ComponentFromFactory />
        </View>
      );
    };

    const TestComponentWithFactory = withComponentFactory<TestComponentProps>(TestComponent);

    const c = renderer.create(
      <ComponentFactoryReactContext.Provider value={componentFactory}>
        <TestComponentWithFactory customProperty="yyy" />
      </ComponentFactoryReactContext.Provider>
    );

    expect(c).toMatchSnapshot();
  });

  test('should pass factory to wrapped component from props', () => {
    const propsFactoryComponents = new Map();
    const contextFactoryComponents = new Map();

    type TestComponentProps = {
      customProperty?: string;
      componentFactory?: ComponentFactory | null;
    }

    contextFactoryComponents.set('xxx', () => <Text>I'm component from context factory</Text>);

    propsFactoryComponents.set('xxx', () => <Text>I'm component from props factory</Text>);

    const propsComponentFactory: ComponentFactory = (componentName) =>
      propsFactoryComponents.get(componentName);
    const contextComponentFactory: ComponentFactory = (componentName) =>
      contextFactoryComponents.get(componentName);

    const TestComponent = (props: TestComponentProps) => {
      if (!props.componentFactory) return null;
      
      const ComponentFromFactory = props.componentFactory('xxx') as React.ComponentType;

      return (
        <View>
          <Text>Hello world...</Text>
          <Text>Custom property:{props.customProperty}</Text>
          <ComponentFromFactory />
        </View>
      );
    };

    const TestComponentWithFactory = withComponentFactory<TestComponentProps>(TestComponent);

    const c = renderer.create(
      <ComponentFactoryReactContext.Provider value={contextComponentFactory}>
        <TestComponentWithFactory componentFactory={propsComponentFactory} customProperty="yyy" />
      </ComponentFactoryReactContext.Provider>
    );

    expect(c).toMatchSnapshot();
  });
});
