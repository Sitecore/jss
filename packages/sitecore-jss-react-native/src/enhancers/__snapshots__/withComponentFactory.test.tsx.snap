// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withComponentFactory() should pass factory to wrapped component from context 1`] = `
<View>
  <Text>
    Hello world...
  </Text>
  <Text>
    Custom property:
    yyy
  </Text>
  <Text>
    I'm component from factory
  </Text>
</View>
`;

exports[`withComponentFactory() should pass factory to wrapped component from props 1`] = `
<View>
  <Text>
    Hello world...
  </Text>
  <Text>
    Custom property:
    yyy
  </Text>
  <Text>
    I'm component from props factory
  </Text>
</View>
`;
